pulp_rpm/tests/functional/api/test_retention_policy.py

```diff
@@ -0,0 +1,172 @@
+           version only).
+        5. Sync the remote one more time.
+        6. Assert that repository version is different from the previous one.
+        7. Assert the repository version we end with has only one version of each package.
+        """
+        repo = self.repo_api.create(gen_repo())
+        self.addCleanup(self.repo_api.delete, repo.pulp_href)
+
+        remote = self.remote_api.create(gen_rpm_remote())
+        self.addCleanup(self.remote_api.delete, remote.pulp_href)
+
+        self.sync(repository=repo, remote=remote, optimize=False)
+        repo = self.repo_api.read(repo.pulp_href)
+
+        # Test that, by default, everything is retained / nothing is tossed out.
+        self.assertDictEqual(get_content_summary(repo.to_dict()), RPM_FIXTURE_SUMMARY)
+        self.assertDictEqual(get_added_content_summary(repo.to_dict()), RPM_FIXTURE_SUMMARY)
+
+        # Set the retention policy to retain only 1 version of each package
+        repo_data = repo.to_dict()
+        repo_data.update({'retain_package_versions': 1})
+        self.repo_api.update(repo.pulp_href, data=repo_data)
+        repo = self.repo_api.read(repo.pulp_href)
+
+        self.sync(repository=repo, remote=remote, optimize=False)
+        repo = self.repo_api.read(repo.pulp_href)
+
+        # Test that only one version of each package is present
+        self.assertTrue(
+            self.check_retention_policy(get_content(repo.to_dict())[PULP_TYPE_PACKAGE], 1)
+        )
+        # Test that (only) 4 RPMs were removed (no advisories etc. touched)
+        self.assertDictEqual(
+            get_removed_content_summary(repo.to_dict()),
+            {PULP_TYPE_PACKAGE: 4}
+        )
+        # Test that the versions that were removed are the versions we expect.
+        versions_for_packages = self.versions_for_packages(
+            get_removed_content(repo.to_dict())[PULP_TYPE_PACKAGE]
+        )
+        self.assertDictEqual(
+            versions_for_packages,
+            {'duck': ['0.7', '0.6'], 'kangaroo': ['0.2'], 'walrus': ['0.71']},
+            versions_for_packages
+        )
+        # TODO: Test that modular RPMs unaffected?
+
+    def test_mirror_sync_with_retention_fails(self):
+        """Verify functionality with sync.
+
+        Do the following:
+
+        1. Create a repository with 'retain_package_versions' set, and a remote.
+        2. Sync the remote in mirror mode.
+        3. Assert that the sync fails.
+        """
+        repo_data = gen_repo()
+        repo_data.update({'retain_package_versions': 1})
+        repo = self.repo_api.create(repo_data)
+        self.addCleanup(self.repo_api.delete, repo.pulp_href)
+
+        remote = self.remote_api.create(gen_rpm_remote())
+        self.addCleanup(self.remote_api.delete, remote.pulp_href)
+
+        with self.assertRaises(ApiException):
+            self.sync(repository=repo, remote=remote, optimize=False, mirror=True)
```

Comment:
Not sure if there's a way to assert the specific code (400).

Reply:
There is:
https://github.com/pulp/pulp_container/blob/c91bed1868a1c3c536f90637d43b0be74d50c573/pulp_container/tests/functional/api/test_crud_distributions.py#L163